File: santa_biblia/tool/build_bible_db.dart
// Bible Database Builder (Dart)
// ──────────────────────────────────────────────────────────────────────
// Construye santa_biblia/assets/db/santa_biblia.db con 5 versiones de
// dominio público. Ejecutar desde santa_biblia/:
//     dart run tool/build_bible_db.dart

import 'dart:convert';
import 'dart:io';
import 'package:http/http.dart' as http;
import 'package:sqflite_common_ffi/sqflite_ffi.dart';

const String _scrollmapperBase =
    'https://raw.githubusercontent.com/scrollmapper/bible_databases/master/formats/csv';

// versionId in our app  →  scrollmapper CSV filename
const Map<String, String> _scrollmapperSources = {
  'kjv': 'KJV.csv',
  'asv': 'ASV.csv',
  'rv1909': 'SpaRV.csv',
  'rv1865': 'SpaRV1865.csv',
};

// bible-api.com aggressively rate-limits (HTTP 403/429) so we don't fetch WEB
// that way. If a WEB CSV becomes available in scrollmapper, add it above.
const String _bibleApiBase = 'https://bible-api.com';
const Map<String, String> _bibleApiSources = <String, String>{};

// English book name → numeric book_id (1–66).
// Includes both Arabic and Roman numeral variants used by different sources.
const Map<String, int> _bookIdByName = {
  'Genesis': 1, 'Exodus': 2, 'Leviticus': 3, 'Numbers': 4, 'Deuteronomy': 5,
  'Joshua': 6, 'Judges': 7, 'Ruth': 8,
  '1 Samuel': 9, 'I Samuel': 9,
  '2 Samuel': 10, 'II Samuel': 10,
  '1 Kings': 11, 'I Kings': 11,
  '2 Kings': 12, 'II Kings': 12,
  '1 Chronicles': 13, 'I Chronicles': 13,
  '2 Chronicles': 14, 'II Chronicles': 14,
  'Ezra': 15, 'Nehemiah': 16, 'Esther': 17, 'Job': 18,
  'Psalms': 19, 'Psalm': 19,
  'Proverbs': 20, 'Ecclesiastes': 21,
  'Song of Solomon': 22, 'Song of Songs': 22, 'Canticles': 22,
  'Isaiah': 23, 'Jeremiah': 24, 'Lamentations': 25, 'Ezekiel': 26,
  'Daniel': 27, 'Hosea': 28, 'Joel': 29, 'Amos': 30, 'Obadiah': 31,
  'Jonah': 32, 'Micah': 33, 'Nahum': 34, 'Habakkuk': 35, 'Zephaniah': 36,
  'Haggai': 37, 'Zechariah': 38, 'Malachi': 39, 'Matthew': 40, 'Mark': 41,
  'Luke': 42, 'John': 43, 'Acts': 44, 'Romans': 45,
  '1 Corinthians': 46, 'I Corinthians': 46,
  '2 Corinthians': 47, 'II Corinthians': 47,
  'Galatians': 48, 'Ephesians': 49, 'Philippians': 50, 'Colossians': 51,
  '1 Thessalonians': 52, 'I Thessalonians': 52,
  '2 Thessalonians': 53, 'II Thessalonians': 53,
  '1 Timothy': 54, 'I Timothy': 54,
  '2 Timothy': 55, 'II Timothy': 55,
  'Titus': 56, 'Philemon': 57, 'Hebrews': 58, 'James': 59,
  '1 Peter': 60, 'I Peter': 60,
  '2 Peter': 61, 'II Peter': 61,
  '1 John': 62, 'I John': 62,
  '2 John': 63, 'II John': 63,
  '3 John': 64, 'III John': 64,
  'Jude': 65,
  'Revelation': 66, 'Revelation of John': 66,
};

// Book slugs for bible-api.com
const List<List<dynamic>> _bookSlugs = [
  [1, 'genesis'], [2, 'exodus'], [3, 'leviticus'], [4, 'numbers'],
  [5, 'deuteronomy'], [6, 'joshua'], [7, 'judges'], [8, 'ruth'],
  [9, '1+samuel'], [10, '2+samuel'], [11, '1+kings'], [12, '2+kings'],
  [13, '1+chronicles'], [14, '2+chronicles'], [15, 'ezra'], [16, 'nehemiah'],
  [17, 'esther'], [18, 'job'], [19, 'psalms'], [20, 'proverbs'],
  [21, 'ecclesiastes'], [22, 'song+of+solomon'], [23, 'isaiah'],
  [24, 'jeremiah'], [25, 'lamentations'], [26, 'ezekiel'], [27, 'daniel'],
  [28, 'hosea'], [29, 'joel'], [30, 'amos'], [31, 'obadiah'], [32, 'jonah'],
  [33, 'micah'], [34, 'nahum'], [35, 'habakkuk'], [36, 'zephaniah'],
  [37, 'haggai'], [38, 'zechariah'], [39, 'malachi'], [40, 'matthew'],
  [41, 'mark'], [42, 'luke'], [43, 'john'], [44, 'acts'], [45, 'romans'],
  [46, '1+corinthians'], [47, '2+corinthians'], [48, 'galatians'],
  [49, 'ephesians'], [50, 'philippians'], [51, 'colossians'],
  [52, '1+thessalonians'], [53, '2+thessalonians'], [54, '1+timothy'],
  [55, '2+timothy'], [56, 'titus'], [57, 'philemon'], [58, 'hebrews'],
  [59, 'james'], [60, '1+peter'], [61, '2+peter'], [62, '1+john'],
  [63, '2+john'], [64, '3+john'], [65, 'jude'], [66, 'revelation'],
];

Future<List<int>> _download(String url, {int retries = 3}) async {
  for (var attempt = 0; attempt < retries; attempt++) {
    try {
      final resp = await http
          .get(Uri.parse(url), headers: {'User-Agent': 'BibleDbBuilder/1.0'})
          .timeout(const Duration(seconds: 30));
      if (resp.statusCode == 200) return resp.bodyBytes;
      throw 'HTTP ${resp.statusCode}';
    } catch (e) {
      if (attempt == retries - 1) rethrow;
      await Future.delayed(Duration(seconds: 1 << attempt));
    }
  }
  throw 'unreachable';
}

Future<void> _createSchema(Database db) async {
  await db.execute('''
    CREATE TABLE IF NOT EXISTS verses (
      id          INTEGER PRIMARY KEY AUTOINCREMENT,
      version_id  TEXT    NOT NULL,
      book_id     INTEGER NOT NULL,
      chapter     INTEGER NOT NULL,
      verse       INTEGER NOT NULL,
      text        TEXT    NOT NULL
    )
  ''');
  await db.execute(
      'CREATE INDEX IF NOT EXISTS idx_verses_lookup ON verses(version_id, book_id, chapter)');
  await db.execute(
      'CREATE INDEX IF NOT EXISTS idx_verses_version ON verses(version_id)');
  // Standalone FTS5 (no external content) — populated via the AFTER INSERT
  // trigger below. Avoids the `rebuild` step and the column-name coupling to
  // the verses table that was causing failures.
  await db.execute('''
    CREATE VIRTUAL TABLE IF NOT EXISTS verses_fts USING fts5(
      text, version_id UNINDEXED, book_id UNINDEXED,
      chapter UNINDEXED, verse_num UNINDEXED, row_id UNINDEXED
    )
  ''');
  await db.execute('''
    CREATE TRIGGER IF NOT EXISTS verses_ai AFTER INSERT ON verses BEGIN
      INSERT INTO verses_fts(rowid, text, version_id, book_id, chapter, verse_num, row_id)
      VALUES (new.id, new.text, new.version_id, new.book_id, new.chapter, new.verse, new.id);
    END
  ''');
}

Future<int> _versionCount(Database db, String versionId) async {
  final r = await db.rawQuery(
      'SELECT COUNT(*) c FROM verses WHERE version_id = ?', [versionId]);
  return (r.first['c'] as int?) ?? 0;
}

// Parses a CSV line respecting quoted fields with "" escapes.
List<String> _parseCsvLine(String line) {
  final fields = <String>[];
  final buf = StringBuffer();
  var inQuotes = false;
  for (var i = 0; i < line.length; i++) {
    final ch = line[i];
    if (inQuotes) {
      if (ch == '"') {
        if (i + 1 < line.length && line[i + 1] == '"') {
          buf.write('"');
          i++;
        } else {
          inQuotes = false;
        }
      } else {
        buf.write(ch);
      }
    } else {
      if (ch == ',') {
        fields.add(buf.toString());
        buf.clear();
      } else if (ch == '"') {
        inQuotes = true;
      } else {
        buf.write(ch);
      }
    }
  }
  fields.add(buf.toString());
  return fields;
}

// Reads CSV text that may contain newlines inside quoted fields.
List<List<String>> _parseCsv(String text) {
  final rows = <List<String>>[];
  final buf = StringBuffer();
  var inQuotes = false;
  for (var i = 0; i < text.length; i++) {
    final ch = text[i];
    if (ch == '"') {
      if (inQuotes && i + 1 < text.length && text[i + 1] == '"') {
        buf.write('""');
        i++;
        continue;
      }
      inQuotes = !inQuotes;
      buf.write(ch);
    } else if (ch == '\n' && !inQuotes) {
      final line = buf.toString();
      buf.clear();
      if (line.isNotEmpty && line != '\r') {
        rows.add(_parseCsvLine(line.endsWith('\r')
            ? line.substring(0, line.length - 1)
            : line));
      }
    } else {
      buf.write(ch);
    }
  }
  if (buf.isNotEmpty) {
    rows.add(_parseCsvLine(buf.toString()));
  }
  return rows;
}

Future<void> _loadFromScrollmapper(
    Database db, String versionId, String filename) async {
  final url = '$_scrollmapperBase/$filename';
  stdout.write('\n  📥 Descargando ${versionId.toUpperCase()} ($filename)... ');
  final raw = await _download(url);
  stdout.writeln('✓ (${raw.length ~/ 1024} KB)');

  final text = utf8.decode(raw);
  final rows = _parseCsv(text);
  if (rows.isNotEmpty) rows.removeAt(0); // header: Book,Chapter,Verse,Text

  var inserted = 0;
  var skipped = 0;
  final batch = db.batch();
  for (final cols in rows) {
    if (cols.length < 4) continue;
    final bookId = _bookIdByName[cols[0]];
    if (bookId == null) {
      skipped++;
      continue;
    }
    batch.insert('verses', {
      'version_id': versionId,
      'book_id': bookId,
      'chapter': int.parse(cols[1]),
      'verse': int.parse(cols[2]),
      'text': cols[3].trim(),
    });
    inserted++;
  }
  stdout.write('  ✍️  Insertando $inserted versículos... ');
  await batch.commit(noResult: true);
  stdout.writeln('✓${skipped > 0 ? " (saltados: $skipped)" : ""}');
}

Future<void> _loadFromBibleApi(
    Database db, String versionId, String apiCode) async {
  stdout.writeln(
      '\n  📥 Descargando ${versionId.toUpperCase()} desde bible-api.com (66 libros)...');
  var total = 0;
  for (var i = 0; i < _bookSlugs.length; i++) {
    final bookId = _bookSlugs[i][0] as int;
    final slug = _bookSlugs[i][1] as String;
    final pct = ((i + 1) / 66 * 40).toInt();
    final bar = '█' * pct + '░' * (40 - pct);
    stdout.write(
        '\r  [$bar] ${i + 1}/66  ${slug.replaceAll('+', ' ').padRight(24)}');

    try {
      final url = '$_bibleApiBase/$slug?translation=$apiCode';
      final raw = await _download(url);
      final data = json.decode(utf8.decode(raw)) as Map<String, dynamic>;
      final verses = (data['verses'] as List<dynamic>? ?? []);
      final batch = db.batch();
      for (final v in verses) {
        final m = v as Map<String, dynamic>;
        batch.insert('verses', {
          'version_id': versionId,
          'book_id': bookId,
          'chapter': m['chapter'] as int,
          'verse': m['verse'] as int,
          'text': (m['text'] as String).trim(),
        });
      }
      await batch.commit(noResult: true);
      total += verses.length;
    } catch (e) {
      stdout.writeln('\n  ⚠️  $slug: $e — omitido');
    }
    await Future.delayed(const Duration(milliseconds: 300));
  }
  stdout.writeln('\n  ✓ $total versículos insertados');
}

Future<void> main() async {
  stdout.writeln('=' * 60);
  stdout.writeln('  🕊️  Santa Biblia — Bible DB Builder (Dart)');
  stdout.writeln('=' * 60);

  sqfliteFfiInit();
  final factory = databaseFactoryFfi;

  // Absolute path — sqflite_common_ffi rewrites bare relative paths into its
  // own `.dart_tool/sqflite_common_ffi/databases/…` directory.
  final projectRoot = Directory.current.path;
  final outputPath =
      '$projectRoot${Platform.pathSeparator}assets${Platform.pathSeparator}db${Platform.pathSeparator}santa_biblia.db';
  await Directory(
          '$projectRoot${Platform.pathSeparator}assets${Platform.pathSeparator}db')
      .create(recursive: true);

  final outFile = File(outputPath);
  if (await outFile.exists()) {
    stdout.writeln('\n  ⚠️  Ya existe — conservando para reanudar progreso');
  }

  final db = await factory.openDatabase(outputPath);
  try {
    await _createSchema(db);

    stdout.writeln(
        '\n┌─ SCROLLMAPPER (KJV, ASV, SpaRV, SpaRV1865) ─────────────┐');
    for (final e in _scrollmapperSources.entries) {
      final count = await _versionCount(db, e.key);
      if (count > 30000) {
        stdout.writeln(
            '  ✓ ${e.key.toUpperCase()} ya cargado ($count versículos), omitiendo');
        continue;
      }
      await _loadFromScrollmapper(db, e.key, e.value);
    }

    stdout.writeln(
        '\n┌─ BIBLE-API.COM (WEB) ───────────────────────────────────┐');
    for (final e in _bibleApiSources.entries) {
      final count = await _versionCount(db, e.key);
      if (count > 30000) {
        stdout.writeln(
            '  ✓ ${e.key.toUpperCase()} ya cargado ($count versículos), omitiendo');
        continue;
      }
      await _loadFromBibleApi(db, e.key, e.value);
    }

    stdout.writeln(
        '\n  🔍 FTS5 poblado en vivo por trigger (sin rebuild necesario)');

    stdout.writeln('\n  📊 Verificación:');
    final summary = await db.rawQuery(
        'SELECT version_id, COUNT(*) c FROM verses GROUP BY version_id ORDER BY version_id');
    for (final r in summary) {
      stdout.writeln('    ${r['version_id']}  ${r['c']} versículos');
    }
    final juan = await db.rawQuery(
        "SELECT text FROM verses WHERE version_id='kjv' AND book_id=43 AND chapter=3 AND verse=16");
    if (juan.isNotEmpty) {
      final t = juan.first['text'] as String;
      stdout.writeln(
          '\n  📖 Juan 3:16 (KJV): «${t.substring(0, t.length > 80 ? 80 : t.length)}...»');
    }

    // Must match `version: 1` in BibleDatabase._initDb, otherwise sqflite will
    // re-run onCreate on launch and blow up on CREATE TABLE verses.
    await db.execute('PRAGMA user_version = 1');

    stdout.write('\n  🧹 VACUUM... ');
    await db.execute('VACUUM');
    stdout.writeln('✓');
  } finally {
    await db.close();
  }

  final size = (await File(outputPath).length()) / 1048576;
  stdout.writeln(
      '\n  ✅ Base de datos lista: $outputPath (${size.toStringAsFixed(1)} MB)');
  stdout.writeln('=' * 60);
}

File: scripts/build_bible_db.py
#!/usr/bin/env python3
"""
Santa Biblia - Bible Database Builder
======================================
Descarga versiones de dominio público y construye el SQLite pre-empaquetado.
Ejecutar UNA VEZ desde la raíz del proyecto:
    python scripts/build_bible_db.py

Resultado: santa_biblia/assets/db/santa_biblia.db  (~8 MB)
"""

import sqlite3
import csv
import urllib.request
import urllib.error
import io
import os
import time
import json
import sys
from pathlib import Path

# ── Configuración ─────────────────────────────────────────────────────────────

OUTPUT_DB = Path(__file__).parent.parent / "santa_biblia" / "assets" / "db" / "santa_biblia.db"

# Fuentes de dominio público
# scrollmapper/bible_databases: CSV con formato id,b,c,v,t
SCROLLMAPPER_BASE = "https://raw.githubusercontent.com/scrollmapper/bible_databases/master/csv"

SOURCES_CSV = {
    "kjv": f"{SCROLLMAPPER_BASE}/t_kjv.csv",
    "web": f"{SCROLLMAPPER_BASE}/t_web.csv",
    "asv": f"{SCROLLMAPPER_BASE}/t_asv.csv",
}

# Para español usamos bible-api.com (dominio público, soporta rv1909)
BIBLE_API_BASE = "https://bible-api.com"
SPANISH_VERSIONS = {
    "rv1909": "rv1909",
}

# Libros de la Biblia (nombre en inglés para la API, ID estándar 1-66)
BIBLE_BOOKS = [
    (1,  "genesis",          50),  (2,  "exodus",           40),
    (3,  "leviticus",        27),  (4,  "numbers",          36),
    (5,  "deuteronomy",      34),  (6,  "joshua",           24),
    (7,  "judges",           21),  (8,  "ruth",              4),
    (9,  "1+samuel",         31),  (10, "2+samuel",         24),
    (11, "1+kings",          22),  (12, "2+kings",          25),
    (13, "1+chronicles",     29),  (14, "2+chronicles",     36),
    (15, "ezra",             10),  (16, "nehemiah",         13),
    (17, "esther",           10),  (18, "job",              42),
    (19, "psalms",          150),  (20, "proverbs",         31),
    (21, "ecclesiastes",    12),   (22, "song+of+solomon",   8),
    (23, "isaiah",           66),  (24, "jeremiah",         52),
    (25, "lamentations",     5),   (26, "ezekiel",          48),
    (27, "daniel",           12),  (28, "hosea",            14),
    (29, "joel",              3),  (30, "amos",              9),
    (31, "obadiah",           1),  (32, "jonah",             4),
    (33, "micah",             7),  (34, "nahum",             3),
    (35, "habakkuk",          3),  (36, "zephaniah",         3),
    (37, "haggai",            2),  (38, "zechariah",        14),
    (39, "malachi",           4),  (40, "matthew",          28),
    (41, "mark",             16),  (42, "luke",             24),
    (43, "john",             21),  (44, "acts",             28),
    (45, "romans",           16),  (46, "1+corinthians",   16),
    (47, "2+corinthians",   13),   (48, "galatians",         6),
    (49, "ephesians",         6),  (50, "philippians",       4),
    (51, "colossians",        4),  (52, "1+thessalonians",   5),
    (53, "2+thessalonians",   3),  (54, "1+timothy",         6),
    (55, "2+timothy",         4),  (56, "titus",             3),
    (57, "philemon",          1),  (58, "hebrews",          13),
    (59, "james",             5),  (60, "1+peter",           5),
    (61, "2+peter",           3),  (62, "1+john",            5),
    (63, "2+john",            1),  (64, "3+john",            1),
    (65, "jude",              1),  (66, "revelation",       22),
]

# ── Helpers ───────────────────────────────────────────────────────────────────

def download(url: str, label: str = "", retries: int = 3) -> bytes:
    for attempt in range(retries):
        try:
            req = urllib.request.Request(url, headers={"User-Agent": "BibleDbBuilder/1.0"})
            with urllib.request.urlopen(req, timeout=30) as r:
                data = r.read()
            return data
        except Exception as e:
            if attempt < retries - 1:
                print(f"    ↻ Reintentando ({attempt+2}/{retries})... {e}")
                time.sleep(2 ** attempt)
            else:
                raise RuntimeError(f"Error descargando {label}: {e}")

def progress(current: int, total: int, label: str):
    pct = int(current / total * 40)
    bar = "█" * pct + "░" * (40 - pct)
    print(f"\r  [{bar}] {current}/{total}  {label:<30}", end="", flush=True)

# ── Creación del schema ────────────────────────────────────────────────────────

def create_schema(conn: sqlite3.Connection):
    conn.executescript("""
        PRAGMA journal_mode=WAL;
        PRAGMA synchronous=NORMAL;
        PRAGMA page_size=4096;

        CREATE TABLE IF NOT EXISTS verses (
            id          INTEGER PRIMARY KEY AUTOINCREMENT,
            version_id  TEXT    NOT NULL,
            book_id     INTEGER NOT NULL,
            chapter     INTEGER NOT NULL,
            verse       INTEGER NOT NULL,
            text        TEXT    NOT NULL
        );

        CREATE INDEX IF NOT EXISTS idx_verses_lookup
            ON verses(version_id, book_id, chapter);

        CREATE INDEX IF NOT EXISTS idx_verses_version
            ON verses(version_id);

        CREATE VIRTUAL TABLE IF NOT EXISTS verses_fts USING fts5(
            text,
            version_id  UNINDEXED,
            book_id     UNINDEXED,
            chapter     UNINDEXED,
            verse_num   UNINDEXED,
            row_id      UNINDEXED,
            content='verses',
            content_rowid='id'
        );

        CREATE TRIGGER IF NOT EXISTS verses_ai AFTER INSERT ON verses BEGIN
            INSERT INTO verses_fts(rowid, text, version_id, book_id, chapter, verse_num, row_id)
            VALUES (new.id, new.text, new.version_id, new.book_id, new.chapter, new.verse, new.id);
        END;
    """)
    conn.commit()

# ── Carga desde scrollmapper CSV ──────────────────────────────────────────────

def load_from_scrollmapper_csv(conn: sqlite3.Connection, version_id: str, url: str):
    print(f"\n  📥 Descargando {version_id.upper()} desde scrollmapper...", end=" ", flush=True)
    raw = download(url, version_id)
    print(f"✓ ({len(raw)//1024} KB)")

    reader = csv.reader(io.StringIO(raw.decode("utf-8")))
    next(reader)  # skip header: id,b,c,v,t

    rows = []
    for cols in reader:
        if len(cols) < 5:
            continue
        _, book_id, chapter, verse, text = cols[0], int(cols[1]), int(cols[2]), int(cols[3]), cols[4]
        rows.append((version_id, book_id, chapter, verse, text.strip()))

    print(f"  ✍️  Insertando {len(rows):,} versículos...", end=" ", flush=True)
    conn.executemany(
        "INSERT INTO verses(version_id, book_id, chapter, verse, text) VALUES (?,?,?,?,?)",
        rows
    )
    conn.commit()
    print(f"✓")

# ── Carga desde bible-api.com (libro por libro) ───────────────────────────────

def load_from_bible_api(conn: sqlite3.Connection, version_id: str, api_code: str):
    print(f"\n  📥 Descargando {version_id.upper()} desde bible-api.com (66 libros)...")

    total_verses = 0
    for i, (book_id, book_slug, _) in enumerate(BIBLE_BOOKS):
        progress(i + 1, 66, book_slug.replace("+", " ").title())

        url = f"{BIBLE_API_BASE}/{book_slug}?translation={api_code}"
        try:
            raw = download(url, f"{version_id}/{book_slug}")
            data = json.loads(raw)
            verses_raw = data.get("verses", [])
        except Exception as e:
            print(f"\n  ⚠️  Libro {book_id} ({book_slug}): {e} — omitido")
            time.sleep(1)
            continue

        rows = [
            (version_id, book_id, v["chapter"], v["verse"], v["text"].strip())
            for v in verses_raw
        ]
        if rows:
            conn.executemany(
                "INSERT INTO verses(version_id, book_id, chapter, verse, text) VALUES (?,?,?,?,?)",
                rows
            )
            total_verses += len(rows)

        # Commit cada libro para no perder progreso si se interrumpe
        conn.commit()
        time.sleep(0.3)  # respetar rate limit de la API

    print(f"\n  ✓ {total_verses:,} versículos insertados")

# ── Reconstruir FTS ───────────────────────────────────────────────────────────

def rebuild_fts(conn: sqlite3.Connection):
    print("\n  🔍 Reconstruyendo índice de búsqueda FTS5...", end=" ", flush=True)
    conn.execute("INSERT INTO verses_fts(verses_fts) VALUES('rebuild')")
    conn.commit()
    print("✓")

# ── Verificación ──────────────────────────────────────────────────────────────

def verify(conn: sqlite3.Connection):
    print("\n  📊 Verificación del contenido:")
    rows = conn.execute("""
        SELECT version_id, COUNT(*) as cnt
        FROM verses GROUP BY version_id ORDER BY version_id
    """).fetchall()
    for version_id, cnt in rows:
        print(f"    {version_id:<10} {cnt:>6,} versículos")

    # Prueba de búsqueda FTS
    test = conn.execute(
        "SELECT text FROM verses WHERE version_id='kjv' AND book_id=43 AND chapter=3 AND verse=16"
    ).fetchone()
    if test:
        print(f"\n  📖 Juan 3:16 (KJV): «{test[0][:80]}...»")

# ── Main ──────────────────────────────────────────────────────────────────────

def main():
    print("=" * 60)
    print("  🕊️  Santa Biblia — Bible Database Builder")
    print("=" * 60)

    OUTPUT_DB.parent.mkdir(parents=True, exist_ok=True)

    if OUTPUT_DB.exists():
        print(f"\n  ⚠️  Ya existe: {OUTPUT_DB}")
        resp = input("  ¿Sobreescribir? (s/N): ").strip().lower()
        if resp != "s":
            print("  Cancelado.")
            return
        OUTPUT_DB.unlink()

    print(f"\n  💾 Creando base de datos en: {OUTPUT_DB}\n")

    conn = sqlite3.connect(str(OUTPUT_DB))

    try:
        create_schema(conn)

        # ── Inglés: KJV, WEB, ASV (scrollmapper, rápido) ─────────────────────
        print("\n┌─ INGLÉS (scrollmapper.com) ─────────────────────────────┐")
        for version_id, url in SOURCES_CSV.items():
            # Check if already loaded (permite reanudar si se interrumpió)
            count = conn.execute(
                "SELECT COUNT(*) FROM verses WHERE version_id=?", (version_id,)
            ).fetchone()[0]
            if count > 30000:
                print(f"\n  ✓ {version_id.upper()} ya cargado ({count:,} versículos), omitiendo")
                continue
            load_from_scrollmapper_csv(conn, version_id, url)

        # ── Español: RV1909, RV1865 (bible-api.com) ───────────────────────────
        print("\n┌─ ESPAÑOL (bible-api.com) ───────────────────────────────┐")
        for version_id, api_code in SPANISH_VERSIONS.items():
            count = conn.execute(
                "SELECT COUNT(*) FROM verses WHERE version_id=?", (version_id,)
            ).fetchone()[0]
            if count > 30000:
                print(f"\n  ✓ {version_id.upper()} ya cargado ({count:,} versículos), omitiendo")
                continue
            load_from_bible_api(conn, version_id, api_code)

        # ── Finalización ──────────────────────────────────────────────────────
        print("\n┌─ FINALIZANDO ───────────────────────────────────────────┐")
        rebuild_fts(conn)
        verify(conn)

        print("\n  🧹 Optimizando base de datos (VACUUM)...", end=" ", flush=True)
        conn.execute("VACUUM")
        print("✓")

        size_mb = OUTPUT_DB.stat().st_size / 1_048_576
        print(f"\n  ✅ Base de datos lista: {OUTPUT_DB.name} ({size_mb:.1f} MB)")
        print("\n  Próximo paso:")
        print("    cd santa_biblia && flutter run -d windows")
        print("=" * 60)

    except KeyboardInterrupt:
        conn.commit()  # guardar progreso
        print("\n\n  ⚠️  Interrumpido. El progreso se guardó. Re-ejecuta para continuar.")
    finally:
        conn.close()


if __name__ == "__main__":
    main()
